presto-main/src/main/java/com/facebook/presto/split/BufferingSplitSource.java

```diff
@@ -56,21 +56,31 @@ public ConnectorTransactionHandle getTransactionHandle()
     public ListenableFuture<List<Split>> getNextBatch(int maxSize)
     {
         checkArgument(maxSize > 0, "Cannot fetch a batch of zero size");
-        List<Split> result = synchronizedList(new ArrayList<>(maxSize));
-        ListenableFuture<?> future = fetchSplits(Math.min(bufferSize, maxSize), maxSize, result);
-        return Futures.transform(future, ignored -> ImmutableList.copyOf(result));
+        return Futures.transform(getNextBatch(NOT_PARTITIONED, maxSize), SplitBatch::getSplits);
     }
 
-    private ListenableFuture<?> fetchSplits(int min, int max, List<Split> output)
+    @Override
+    public ListenableFuture<SplitBatch> getNextBatch(ConnectorPartitionHandle partitionHandle, int maxSize)
+    {
+        checkArgument(maxSize > 0, "Cannot fetch a batch of zero size");
+        FetchSplitsResult fetchSplitsResult = new FetchSplitsResult();
```

Comment:
I modified my code to work in a way similar to the approach you described. But I kept the Futures.transform (instead of extending AbstractFuture) so that it doesn't keep classes that are only needed while trying to complete the future after its completion.